tools/src/prettier_printer.ml

```diff
@@ -0,0 +1,271 @@
+module DSL = struct
+  type namedField = {name: string; value: oak}
+
+  and oak =
+    | Application of string * oak
+    | Record of namedField list
+    | Ident of string
+    | Tuple of namedField list
+    | List of oak list
+    | String of string
```

Comment:
> Just to make sure I understand the justification of introducing another representation - the intention here is to have a generic pretty printer that we can use for debugging, correct? Where tast is the first thing we've made a debug printer for, but we can extend it as needed.

Yes, I wanted to pattern match in `tools.ml` but wasn't sure what shape I was aiming for. So, I created a new representation and wrote a printer for it. In the future, this could be reused for the untyped tree.

I believe the `loc` aspect can be captured as you described. It would be helpful to have a concrete example to ensure I fully understand what you would like to see.

Reply:
Sure! Have a look through this file: https://github.com/rescript-lang/rescript-vscode/blob/master/analysis/src/DumpAst.ml

That prints (parts of) the parsetree, and marks structures with whether they hold the cursor or not (or if the loc is broken).

Reply:
Okay, in the tree traversal you would like to pass the cursor position and have that print something special if a node contains the cursor?

Reply:
Exactly, or if the loc is broken. Reason is that both of those are very important when working with things like hovers, autocomplete, etc. I don't need to see the actual locs I think, just if the cursor is in there and/or if the loc is broken. For actual locs, it's easy enough to print via `bsc`.

Reply:
A couple of questions there:

I'm not sure if I understand what a broken loc is. What does that mean?

> I don't need to see the actual locs 

I'm a little surprised by this. Can you elaborate on this? Or is it more a practical thing?

A while ago you told me that `bsc -dtypedtree` dumps the typed tree, however, inside a real project you need to pass in the dependencies and other flags. Is there an easy way find out what other arguments would be required? Like what is `bunx rescript` eventually gonna pass down?

Reply:
1. Broken loc: A broken loc is what the parser inserts when a syntax error is produced that doesn't have a clear loc by itself. Remember that the parser and editor tooling does most of its work on broken rather than complete sources. So, sometimes the parser inserts things like `%rescript.exprHole` with a "broken" loc at a place where it couldn't figure out more about what type of expression or pattern you're trying to write. These are important cues that help when doing autocomplete.
2. Not needing the actual locs: Exactly, it's a practical thing. They are needed occasionally when debugging something intricate, but mostly what you're after when working on editor features (or debugging existing features) is figuring out where the cursor is located, not what those exact locs are.
3. When you just care about what editor sees/works on, you don't need to know that much about what `bsc` would run when _compiling_. That's when those flags are needed. If you want to see precisely what the _editor_ sees via the parser, this command is good enough (and it requires no knowledge of libs, PPXes etc): `bsc whatever.res -dparsetree -ignore-parse-errors -only-parse -bs-no-builtin-ppx -bs-loc`. This will give you just the parsetree back, just as the parser sees it (which is what the editor operates on). No PPXes (including disabling the builtin PPXes) and `-bs-loc` will give you the actual locs.